In [2]:
import pandas as pd

## Question 1
Make sure that the theme is also included in dataset `sets`:

1. Load the CSV data into dataframes. Be sure to load every file in a separate variable so you can analyze them easily.  
   The URLs for the files reside are as follows, one for each table:
   * `https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/colors.csv`
   * `https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/elements.csv`
   * `https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/inventories.csv`
   * `https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/inventory_minifigs.csv`
   * `https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/inventory_parts.csv`
   * `https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/inventory_sets.csv`
   * `https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/minifigs.csv`
   * `https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/parts.csv`
   * `https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/part_categories.csv`
   * `https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/part_relationships.csv`
   * `https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/sets.csv`
   * `https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/themes_hierachie.csv`
2. Link `themes` to `sets`. The hierarchy is already solved for you in this `themes` set
3. Create a new field (variable, column) that indicates with 0/1 if a set belongs to the Creator theme

When reading all files at once in one script, you might see an error because of throttling at GitHub.
The solution to this it to split the script into multiple cells, each of which load a part of the data.

In [3]:
# Reading data from a CSV file: pd.read_csv
# All tables go into their own separate dataframe and variable
colors = pd.read_csv("https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/colors.csv")
elements = pd.read_csv("https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/elements.csv")
inventories = pd.read_csv("https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/inventories.csv")
inventory_minifigs = pd.read_csv("https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/inventory_minifigs.csv")
inventory_parts = pd.read_csv("https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/inventory_parts.csv")
inventory_sets = pd.read_csv("https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/inventory_sets.csv")
minifigs = pd.read_csv("https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/minifigs.csv")
parts = pd.read_csv("https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/parts.csv")
part_categories = pd.read_csv("https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/part_categories.csv")
part_relationships = pd.read_csv("https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/part_relationships.csv")
sets = pd.read_csv("https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/sets.csv")
themes = pd.read_csv("https://github.com/wortell-smart-learning/lego-casus/raw/main/dataset/themes_hierachie.csv")

In [4]:
# link themes to sets
themes_sets = themes.merge(sets, left_on='id', right_on='theme_id')
themes_sets.head(10) # Display top 10 rows to show the result

id   name_x parent_name set_num                          name_y  year  \
0   1  Technic     Technic   001-1                           Gears  1965   
1   1  Technic     Technic   002-1  4.5V Samsonite Gears Motor Set  1965   
2   1  Technic     Technic  1030-1  TECHNIC I: Simple Machines Set  1985   
3   1  Technic     Technic  1038-1              ERBIE the Robo-Car  1985   
4   1  Technic     Technic  1039-1            Manual Control Set 1  1986   
5   1  Technic     Technic  1168-1                     Battery Box  1986   
6   1  Technic     Technic  1314-1        Stop bush / Small pulley  1987   
7   1  Technic     Technic  1315-1                      Piston Rod  1987   
8   1  Technic     Technic  1316-1                   Connector peg  1987   
9   1  Technic     Technic  1317-1              TECHNIC Chainlinks  1987   

   theme_id  num_parts  
0         1         43  
1         1          3  
2         1        191  
3         1        120  
4         1         39  
5         1          1  
6         1        210  
7         1         50  
8         1        150  
9         1        350

### Before we continue
There are apparently two "theme" columns in the table `themes`:

* name (now renamed to `name_x` to avoid ambiguous column names)
* parent_name

From the data model we know already that this was (once) a parent-child hierarchy - let's do some further analysis:

In [5]:
print("Number of sets with parent_name Creator: ", themes_sets.query('parent_name == "Creator"')["id"].count()) # Count the number of sets with parent_name Creator
print("Number of sets with name_x Creator: ", themes_sets.query('name_x == "Creator"')["id"].count()) # Count the number of sets with name_x Creator
print("Values for parent_name when child_name is Creator: ", themes_sets.query('name_x == "Creator"')["name_x"].value_counts())
print("Values for child name when parent_name is Creator: ", themes_sets.query('parent_name == "Creator"')["name_x"].value_counts())


Number of sets with parent_name Creator:  540
Number of sets with name_x Creator:  124
Values for parent_name when child_name is Creator:  name_x
Creator    124
Name: count, dtype: int64
Values for child name when parent_name is Creator:  name_x
Creator 3-in-1    182
Basic Set         120
Creator            92
Creator Expert     58
Early Creator      29
Supplemental       22
Food & Drink       16
Basic Model         8
Creature            5
Construction        3
Traffic             3
Castle              1
Building            1
Name: count, dtype: int64


Based on this analysis, we can state that:

* When name_x is 'Creator', parent_name is also 'Creator'
* When parent_name is 'Creator', name_x can be different values, for example:
  * Early Creator
  * Creator Expert
  * Creator 3-in-1
  * Basic Set

We can safely assume that all of these still are "Creator" sets, but have some sub-theme.
Therefor, the filter should be on `parent_name`

In [6]:
# Create a new field that indicates with 0/1 if a set belongs to the Creator theme
themes_sets['is_creator'] = (themes_sets['parent_name'] == 'Creator').astype(int)

In [7]:
themes_sets.query('is_creator == 1').head(5) # 5 examples for rows with creator theme

id   name_x parent_name  set_num          name_y  year  theme_id  \
580  22  Creator     Creator  10664-1  Creative Tower  2013        22   
581  22  Creator     Creator  11938-1           Robot  2020        22   
582  22  Creator     Creator  11939-1         Octopus  2020        22   
583  22  Creator     Creator  11940-1        Fortress  2020        22   
584  22  Creator     Creator  11941-1            Frog  2020        22   

     num_parts  is_creator  
580       1600           1  
581         45           1  
582         63           1  
583         52           1  
584         56           1

In [8]:
themes_sets.query('is_creator == 0').head(5) # 5 examples for rows without creator theme

id   name_x parent_name set_num                          name_y  year  \
0   1  Technic     Technic   001-1                           Gears  1965   
1   1  Technic     Technic   002-1  4.5V Samsonite Gears Motor Set  1965   
2   1  Technic     Technic  1030-1  TECHNIC I: Simple Machines Set  1985   
3   1  Technic     Technic  1038-1              ERBIE the Robo-Car  1985   
4   1  Technic     Technic  1039-1            Manual Control Set 1  1986   

   theme_id  num_parts  is_creator  
0         1         43           0  
1         1          3           0  
2         1        191           0  
3         1        120           0  
4         1         39           0

In [9]:
themes_sets.query('is_creator == 0 and parent_name == "Creator"') # Check if there are rows with is_creator == 0 and parent_name == 'Creator'

Empty DataFrame
Columns: [id, name_x, parent_name, set_num, name_y, year, theme_id, num_parts, is_creator]
Index: []

Question: What % of all sets do you own?

In [10]:
num_creator_sets = themes_sets['is_creator'].sum()
print("Number of sets that belong to the Creator theme:", num_creator_sets)
num_total_sets = len(themes_sets)
print("Total number of sets:", num_total_sets)
print(f"Percentage of total sets owned: {num_creator_sets / num_total_sets * 100:.2f}")

Number of sets that belong to the Creator theme: 540
Total number of sets: 17835
Percentage of total sets owned: 3.03


Three things to note:

1. Because the `is_creator` is a number, we can add it to get the number of creator sets
2. I've used a *format string*. Basically the following three expressions yield the same result:
```python
print("Perc of sets owned:", (num_creator_sets / num_total_sets * 100))
print("Perc of sets owned: {placeholder}".format( placeholder=(num_creator_sets / num_total_sets * 100))) # You can use any name instead of placeholder, of cours
print(f"Perc of sets owned: {(num_creator_sets / num_total_sets * 100)}")
```
3. The `:.2f` signifies the number of decimal places that should be displayed in a floating point number. Instead of displaying `3.027754415475189` it now displays `3.03`.


## Question 2:

Filter the dataset based on your created variable **isCreator**

In [11]:
creator_sets = themes_sets.query('is_creator == 1')

Link the data sources together so that you know which parts it contains for all sets together. 

You need the following tables for this:

* Inventories
* Inventory_parts
* Parts
* Colors

This table is unique on `part_num` and `Color_id`

Based on the *rebrickable* datamodel, let's create a join. We will use method chaining in parentheses to create a clear and readable statement for this:

In [12]:
all_my_parts = (creator_sets
    .merge(inventories, left_on='set_num', right_on='set_num', suffixes=('_s', '_i'))
    .merge(inventory_parts, left_on='id_i', right_on='inventory_id', suffixes=(None, '_ip'))
    # .merge(parts, left_on='part_num', right_on='part_num') # I don't think we will need parts for now
    .merge(colors, left_on='color_id', right_on='id', suffixes=(None, '_c'))
)

How many parts do you have in the color "Red"?

In [13]:
# First guess: sum all quantities where the color is "red"
# This is a somewhat dangerous one: there could have been duplication along the way
all_my_parts.query('name == "Red"')["quantity"].sum()

np.int64(26372)

Let's double-check: it should be somewhat in line (but not exactly the same) as the number of parts from `creator_sets`. Let's check this:

In [14]:
print("All my parts:", all_my_parts["quantity"].sum())
print("On set level:", creator_sets["num_parts"].sum())

All my parts: 280639
On set level: 255766


That sounds reasonable, albeit a little bit high:

* Lego puts some spare parts in every box, so it should be higher
* However, we have 540 sets, and there are 24.873 parts "spare" here
* Lego should then have added approximately 46 spare parts in every box, which is kinda high

Let's double check if no unintended duplication is going on:

In [15]:
len(creator_sets)

540

In [16]:
len(creator_sets
    .merge(inventories, left_on='set_num', right_on='set_num', suffixes=('_s', '_i'))
)

564

In [17]:
len(creator_sets
    .merge(inventories, left_on='set_num', right_on='set_num', suffixes=('_s', '_i'))
    .merge(inventory_parts, left_on='id_i', right_on='inventory_id', suffixes=(None, '_ip'))
)

58206

In [18]:
len(creator_sets
    .merge(inventories, left_on='set_num', right_on='set_num', suffixes=('_s', '_i'))
    .merge(inventory_parts, left_on='id_i', right_on='inventory_id', suffixes=(None, '_ip'))
        # .merge(parts, left_on='part_num', right_on='part_num') # I don't think we will need parts for now
    .merge(colors, left_on='color_id', right_on='id', suffixes=(None, '_c'))
)

58206

Apparently, nothing goes wrong: 

* The number of rows increases predictably
* There are a few sets with multiple inventories (540 -> 564)
* There seem to be approximately 100 inventory_parts rows per set (564 -> 58K)
* Merging the colors in doesn't make any difference.

From a high level, nothing is wrong. 

However, we could have a separate look:

* When there are multiple *inventories* associated with one *set*, are lists of parts maybe reported twice?

In [19]:
(creator_sets
    .merge(inventories, left_on='set_num', right_on='set_num', suffixes=('_s', '_i'))
)

id_s         name_x parent_name     set_num  \
0      22        Creator     Creator     10664-1   
1      22        Creator     Creator     11938-1   
2      22        Creator     Creator     11939-1   
3      22        Creator     Creator     11940-1   
4      22        Creator     Creator     11941-1   
..    ...            ...         ...         ...   
559   674  Early Creator     Creator      4906-1   
560   674  Early Creator     Creator      5370-1   
561   674  Early Creator     Creator      7830-1   
562   674  Early Creator     Creator     K4103-1   
563   674  Early Creator     Creator  kabcreat-1   

                                                name_y  year  theme_id  \
0                                       Creative Tower  2013        22   
1                                                Robot  2020        22   
2                                              Octopus  2020        22   
3                                             Fortress  2020        22   
4                                                 Frog  2020        22   
..                                                 ...   ...       ...   
559                                         Helicopter  2005       674   
560  Large Make and Create Bucket with Special LEGO...  2005       674   
561                                  Small Blue Bucket  2002       674   
562   Creator Bucket bundled with 4782 (TRU Exclusive)  2005       674   
563                                     Creator 4-Pack  2001       674   

     num_parts  is_creator   id_i  version  
0         1600           1  12223        1  
1           45           1  69864        1  
2           63           1  72642        1  
3           52           1  73478        1  
4           56           1  76834        1  
..         ...         ...    ...      ...  
559         16           1  12591        1  
560          0           1  10795        1  
561        200           1   6120        1  
562          0           1    104        1  
563          0           1  15080        1  

[564 rows x 11 columns]

A second look makes me wonder at the "version" column. There are *some* sets with a **version** that is equal to 2:

In [20]:
(creator_sets
    .merge(inventories, left_on='set_num', right_on='set_num', suffixes=('_s', '_i'))
)["version"].value_counts()

version
1    539
2     25
Name: count, dtype: int64

In [21]:
(creator_sets
    .merge(inventories, left_on='set_num', right_on='set_num', suffixes=('_s', '_i'))
).query("version == 2")

id_s          name_x parent_name   set_num  \
198    37       Basic Set     Creator    5508-1   
267    48    Supplemental     Creator    6117-1   
282   672  Creator 3-in-1     Creator   31004-1   
289   672  Creator 3-in-1     Creator   31010-1   
293   672  Creator 3-in-1     Creator   31013-1   
296   672  Creator 3-in-1     Creator   31015-1   
350   672  Creator 3-in-1     Creator   31070-1   
365   672  Creator 3-in-1     Creator   31085-1   
390   672  Creator 3-in-1     Creator   31111-1   
405   672  Creator 3-in-1     Creator    4838-1   
417   672  Creator 3-in-1     Creator    4956-1   
421   672  Creator 3-in-1     Creator    4993-1   
440   672  Creator 3-in-1     Creator    5867-1   
445   672  Creator 3-in-1     Creator    5893-1   
449   672  Creator 3-in-1     Creator    6742-1   
461   672  Creator 3-in-1     Creator    6913-1   
474   673  Creator Expert     Creator   10152-3   
479   673  Creator Expert     Creator   10181-1   
486   673  Creator Expert     Creator   10214-1   
490   673  Creator Expert     Creator   10226-1   
498   673  Creator Expert     Creator   10244-1   
502   673  Creator Expert     Creator   10252-1   
506   673  Creator Expert     Creator   10257-1   
508   673  Creator Expert     Creator   10261-1   
530   673  Creator Expert     Creator  HARLEY-1   

                                      name_y  year  theme_id  num_parts  \
198                         Deluxe Brick Box  2010        37        704   
267                        Doors and Windows  2008        48        100   
282                             Fierce Flyer  2013       672        166   
289                                Treehouse  2013       672        356   
293                              Red Thunder  2014       672         66   
296                          Emerald Express  2014       672         56   
350                        Turbo Track Racer  2017       672        670   
365                        Mobile Stunt Show  2018       672        581   
390                              Cyber Drone  2021       672        113   
405                            Mini Vehicles  2008       672         79   
417                                    House  2007       672        731   
421                         Cool Convertible  2008       672        648   
440                          Super Speedster  2010       672        278   
445                            Offroad Power  2010       672       1061   
449                          Mini Off-Roader  2009       672         64   
461                            Blue Roadster  2012       672        152   
474  Maersk Line Container Ship 2006 Edition  2006       673        988   
479                       Eiffel Tower 1:300  2007       673       3428   
486                             Tower Bridge  2010       673       4295   
490                            Sopwith Camel  2012       673        879   
498                         Fairground Mixer  2014       673       1745   
502                        Volkswagen Beetle  2016       673       1167   
506                                 Carousel  2017       673       2670   
508                           Roller Coaster  2018       673       4124   
530          Harley-Davidson Mini Motorcycle  2019       673         71   

     is_creator   id_i  version  
198           1  30080        2  
267           1  74248        2  
282           1  27496        2  
289           1  78334        2  
293           1  45633        2  
296           1  18231        2  
350           1  29830        2  
365           1  95283        2  
390           1  88727        2  
405           1  29359        2  
417           1  42375        2  
421           1  29361        2  
440           1  23778        2  
445           1  23669        2  
449           1  29360        2  
461           1  82798        2  
474           1  33925        2  
479           1  22743        2  
486           1  35420        2  
490           1  20188        2  
498          

Let's zoom in on one with few parts: Emerals Express (`31015-1`):

In [22]:
(creator_sets
    .query("set_num == '31015-1'")
    .merge(inventories, left_on='set_num', right_on='set_num', suffixes=('_s', '_i'))
    .merge(inventory_parts, left_on='id_i', right_on='inventory_id', suffixes=(None, '_ip'))
        # .merge(parts, left_on='part_num', right_on='part_num') # I don't think we will need parts for now
    .merge(colors, left_on='color_id', right_on='id', suffixes=(None, '_c'))
)

id_s          name_x parent_name  set_num           name_y  year  \
0    672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014   
1    672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014   
2    672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014   
3    672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014   
4    672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014   
..   ...             ...         ...      ...              ...   ...   
73   672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014   
74   672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014   
75   672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014   
76   672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014   
77   672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014   

    theme_id  num_parts  is_creator   id_i  version  inventory_id part_num  \
0        672         56           1   3217        1          3217    10201   
1        672         56           1   3217        1          3217    13971   
2        672         56           1   3217        1          3217     2540   
3        672         56           1   3217        1          3217     2817   
4        672         56           1   3217        1          3217     3004   
..       ...        ...         ...    ...      ...           ...      ...   
73       672         56           1  18231        2         18231    87580   
74       672         56           1  18231        2         18231    88930   
75       672         56           1  18231        2         18231    98138   
76       672         56           1  18231        2         18231    98138   
77       672         56           1  18231        2         18231    99780   

    color_id  quantity is_spare  id               name     rgb is_trans  
0          0         1        f   0              Black  05131D        f  
1         71         4        f  71  Light Bluish Gray  A0A5A9        f  
2          0         1        f   0              Black  05131D        f  
3         71         2        f  71  Light Bluish Gray  A0A5A9        f  
4         71         1        f  71  Light Bluish Gray  A0A5A9        f  
..       ...       ...      ...  ..                ...     ...      ...  
73         4         1        f   4                Red  C91A09        f  
74         0         2        f   0              Black  05131D        f  
75        46         1        t  46       Trans-Yellow  F5CD2F        t  
76        46         1        f  46       Trans-Yellow  F5CD2F        t  
77        72         1        f  72   Dark Bluish Gray  6C6E68        f  

[78 rows x 20 columns]

Now the column `is_spare` looks interesting.. Let's focus on only spare parts:

In [23]:
(creator_sets
    .query("set_num == '31015-1'")
    .merge(inventories, left_on='set_num', right_on='set_num', suffixes=('_s', '_i'))
    .merge(inventory_parts, left_on='id_i', right_on='inventory_id', suffixes=(None, '_ip'))
    .query('is_spare == "t"')
        # .merge(parts, left_on='part_num', right_on='part_num') # I don't think we will need parts for now
    .merge(colors, left_on='color_id', right_on='id', suffixes=(None, '_c'))
)

id_s          name_x parent_name  set_num           name_y  year  theme_id  \
0   672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014       672   
1   672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014       672   
2   672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014       672   
3   672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014       672   
4   672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014       672   
5   672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014       672   
6   672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014       672   
7   672  Creator 3-in-1     Creator  31015-1  Emerald Express  2014       672   

   num_parts  is_creator   id_i  version  inventory_id part_num  color_id  \
0         56           1   3217        1          3217    3070b        72   
1         56           1   3217        1          3217     3673        71   
2         56           1   3217        1          3217     6141         0   
3         56           1   3217        1          3217    98138        46   
4         56           1  18231        2         18231    3070b        72   
5         56           1  18231        2         18231     3673        71   
6         56           1  18231        2         18231     6141         0   
7         56           1  18231        2         18231    98138        46   

   quantity is_spare  id               name     rgb is_trans  
0         1        t  72   Dark Bluish Gray  6C6E68        f  
1         1        t  71  Light Bluish Gray  A0A5A9        f  
2         2        t   0              Black  05131D        f  
3         1        t  46       Trans-Yellow  F5CD2F        t  
4         1        t  72   Dark Bluish Gray  6C6E68        f  
5         1        t  71  Light Bluish Gray  A0A5A9        f  
6         2        t   0              Black  05131D        f  
7         1        t  46       Trans-Yellow  F5CD2F        t

It seems we've caught something: there can be two versions of an inventory, so parts can be reported doubly.
What if we use only version == 1?

In [24]:
all_my_parts.query("version == 1")["quantity"].sum()

np.int64(255103)

That number is actually *lower* than the number of parts in the sets. This could well be because the minifigs are counting inside the number of parts of the box, and not in the number of parts of the inventory.
Let's check how many minifig "parts" would be inside

In [25]:
all_my_minifigs = (creator_sets
    .merge(inventories, left_on='set_num', right_on='set_num', suffixes=('_s', '_i'))
    .query("version == 1")
    .merge(inventory_minifigs, left_on='id_i', right_on='inventory_id', suffixes=(None, '_im'))
    # .merge(parts, left_on='part_num', right_on='part_num') # I don't think we will need parts for now
    .merge(minifigs, left_on='fig_num', right_on='fig_num', suffixes=(None, '_m'))
)

In [26]:
(all_my_minifigs["quantity"] * all_my_minifigs["num_parts_m"]).sum()

np.int64(1142)

In [27]:
total_num_parts_including_minifigs = (
    all_my_parts.query("version == 1")["quantity"].sum()
    +
    (all_my_minifigs["quantity"] * all_my_minifigs["num_parts_m"]).sum()
)
total_num_parts_including_minifigs

np.int64(256245)

Which is *a little bit more* than the number of parts on the box:

In [28]:
total_num_parts_according_to_box = creator_sets["num_parts"].sum()
total_num_parts_according_to_box

np.int64(255766)

And it boils down to approximately 0.9 item per box extra, a much more reasonable number:

In [29]:
(total_num_parts_including_minifigs - total_num_parts_according_to_box) / len(creator_sets)

np.float64(0.8870370370370371)

With this new counting method, how many red bricks do we have?

In [30]:
all_my_parts.query("version == 1 and name == 'Red'")["quantity"].sum()

np.int64(25277)

## Question 3
You are also curious how many parts you now have per category. Link your just created table to the table `Parts_Categories` and count the number of parts and sort this table descending.

In [31]:
# Let's first link the table with the parts and parts_categories tables.
# Because of our findings with Q2, we will keep the "version" column in here as well:
my_parts_categories = (
    all_my_parts

    # Note that we are only interested in version 1 here.
    # If you didn't "uncover" the way the versions work, you can also remove this line to see how many parts you would have.
    .query("version == 1")
    .merge(parts, left_on='part_num', right_on='part_num', suffixes=(None, '_p'))
    .merge(part_categories, left_on="part_cat_id", right_on="id", suffixes=(None, "_pc"))
    .merge(colors, left_on="color_id", right_on="id", suffixes=(None, "_c"))
    .rename(columns={"name_pc": "category", "name_p": "part", "name_c": "color"})
)[["part_num", "category", "part", "quantity", "color"]]
my_parts_categories.head(10)

part_num       category                                          part  \
0    2412b  Tiles Special  Tile Special 1 x 2 Grille with Bottom Groove   
1    2412b  Tiles Special  Tile Special 1 x 2 Grille with Bottom Groove   
2     2419  Plates Angled                 Wedge Plate 3 x 6 Cut Corners   
3     2420         Plates                            Plate 2 x 2 Corner   
4     2456         Bricks                                   Brick 2 x 6   
5     2456         Bricks                                   Brick 2 x 6   
6     2456         Bricks                                   Brick 2 x 6   
7     2456         Bricks                                   Brick 2 x 6   
8     2456         Bricks                                   Brick 2 x 6   
9     2456         Bricks                                   Brick 2 x 6   

   quantity              color  
0         4              White  
1         4  Light Bluish Gray  
2         2              Green  
3         2              Black  
4         4              White  
5         4                Red  
6         2                Tan  
7         4               Blue  
8         2              Green  
9         4             Yellow

In [32]:
my_parts_categories[["category", "part", "part_num", "quantity"]].groupby(["category", "part", "part_num"]).sum().sort_values("quantity", ascending=False).head(10)

quantity
category                       part                                 part_num          
Bricks                         Brick 1 x 2                          3004         19301
                               Brick 2 x 2                          3003         16788
                               Brick 1 x 1                          3005         14308
Plates                         Plate 1 x 2                          3023          9189
Bricks                         Brick 2 x 4                          3001          7867
Plates Round Curved and Dishes Plate Round 1 x 1 with Solid Stud    6141          7484
Plates                         Plate 1 x 1                          3024          6511
Bricks                         Brick 1 x 4                          3010          6174
                               Brick 2 x 3                          3002          4838
Bricks Sloped                  Slope 30° 1 x 1 x 2/3 (Cheese Slope) 54200         4484

In [33]:
my_parts_categories[["category", "quantity"]].groupby(["category"]).sum().sort_values("quantity", ascending=False).head(10)

quantity
category                                
Bricks                             82852
Plates                             44867
Bricks Sloped                      21366
Plates Special                     18216
Tiles                              12834
Plates Round Curved and Dishes     10600
Bricks Special                      7806
Bricks Curved                       7665
Bricks Round and Cones              4953
Technic Pins                        4684

What is the average number of parts per creator set?

In [34]:
# First, create a table that lists all different sets, including the number of parts.
my_parts_categories_sets = (
    all_my_parts
    .merge(parts, left_on='part_num', right_on='part_num', suffixes=(None, '_p'))
    # .merge(part_categories, left_on="part_cat_id", right_on="id", suffixes=(None, "_pc"))
    .rename(columns={'name_y': 'set name'})
)[["quantity", "set name", "set_num"]]
qty_per_set = my_parts_categories_sets[['set_num', 'quantity']].groupby('set_num').sum()
qty_per_set.head(10)

quantity
set_num          
10024-1       679
10124-1       677
10152-1       997
10152-2       997
10152-3      1994
10155-1       999
10173-1       958
10177-1      1209
10181-1      6886
10187-1      1642

* First sum the quantity per set (this is important: because we can have multiple tile types per set, each set num will occur multiple times!)
* Then take the average over the total number of tiles

In [35]:
qty_per_set[['quantity']].mean()

quantity    540.73025
dtype: float64

## Question 4

Determine which parts are included in the Hobby Train set (set 10183-1).

In [36]:
# This time, there is only a version 1:
(
sets.query("set_num == '10183-1'")
.merge(inventories, left_on='set_num', right_on='set_num', suffixes=('_s', '_i'))
)

set_num         name  year  theme_id  num_parts    id  version
0  10183-1  Hobby Train  2007       397       1080  5514        1

In [37]:
hobby_train_parts = (
    sets.query("set_num == '10183-1'")
    .merge(inventories, left_on='set_num', right_on='set_num', suffixes=('_s', '_i'))
    .merge(inventory_parts, left_on='id', right_on='inventory_id', suffixes=(None, '_ip'))
)
hobby_train_parts.head(10)

set_num         name  year  theme_id  num_parts    id  version  \
0  10183-1  Hobby Train  2007       397       1080  5514        1   
1  10183-1  Hobby Train  2007       397       1080  5514        1   
2  10183-1  Hobby Train  2007       397       1080  5514        1   
3  10183-1  Hobby Train  2007       397       1080  5514        1   
4  10183-1  Hobby Train  2007       397       1080  5514        1   
5  10183-1  Hobby Train  2007       397       1080  5514        1   
6  10183-1  Hobby Train  2007       397       1080  5514        1   
7  10183-1  Hobby Train  2007       397       1080  5514        1   
8  10183-1  Hobby Train  2007       397       1080  5514        1   
9  10183-1  Hobby Train  2007       397       1080  5514        1   

   inventory_id    part_num  color_id  quantity is_spare  
0          5514       2362b        71        16        f  
1          5514        2377         0         8        f  
2          5514        2377         4         8        f  
3          5514       2412b         0        12        f  
4          5514       2412b         4        12        f  
5          5514        2420         4         4        f  
6          5514        2431         0         4        f  
7          5514        2431         4         6        f  
8          5514  2431pr0028        72         2        f  
9          5514        2432        72         4        f

The number of parts is 1094, which is a little bit more than the 1080 listed on the box (14 extra parts for a 1080-parts box sounds legit)

In [38]:
print(hobby_train_parts["quantity"].sum())


1094


Match this table with the table of all your Creator Collection items. Create a new variable that counts how much you miss of each part.

# First (wrong) guess: EXCEPT ("set operators")
Your first guess (coming from a SQL background might be an EXCEPT query)
Although it is possible to create an EXCEPT functionality in dataframes by creating a boolean dataframe using the `isin` operator (see https://stackoverflow.com/questions/41483823/what-is-the-analogue-of-except-clause-in-sql-in-pandas), it won't work:

Pieces I have (`df_piecesowned`)
| part_num | quantity |
|----------|----------|
| 2377     | 17       |

Pieces I need (`df_piecesneeded`)

| part_num | quantity |
|----------|----------|
| 2377     | 8        |

```python
df_piecesowned = pd.DataFrame({"part_num": [2377], "quantity": [17]})
df_piecesneeded = pd.DataFrame({"part_num": [2377], "quantity": [8]})
df_piecesowned.isin(df_piecesneeded)
```

|   | part_num | quantity |
|---|----------|----------|
| 0 |     True |    False |

This won't tell you anything about how much pieces you're missing

## Second (right) guess: Right merge

Pieces I have (`df_piecesowned`)
| part_num | quantity |
|----------|----------|
| 2377     | 17       |

Pieces I need (`df_piecesneeded`)

| part_num | quantity |
|----------|----------|
| 2377     | 8        |



In [39]:
hobby_train_parts.head(3)

set_num         name  year  theme_id  num_parts    id  version  \
0  10183-1  Hobby Train  2007       397       1080  5514        1   
1  10183-1  Hobby Train  2007       397       1080  5514        1   
2  10183-1  Hobby Train  2007       397       1080  5514        1   

   inventory_id part_num  color_id  quantity is_spare  
0          5514    2362b        71        16        f  
1          5514     2377         0         8        f  
2          5514     2377         4         8        f

In [40]:
my_parts_categories.head(3)

part_num       category                                          part  \
0    2412b  Tiles Special  Tile Special 1 x 2 Grille with Bottom Groove   
1    2412b  Tiles Special  Tile Special 1 x 2 Grille with Bottom Groove   
2     2419  Plates Angled                 Wedge Plate 3 x 6 Cut Corners   

   quantity              color  
0         4              White  
1         4  Light Bluish Gray  
2         2              Green

We need to summarize the parts and categories first:

In [41]:
print(f"Before summarizing: {my_parts_categories.shape}")

my_parts_categories_summary = my_parts_categories[["part_num", "category", "part", "quantity", "color"]].groupby(["part_num", "category", "part", "color"]).sum().reset_index()
print(f"After summarizing: {my_parts_categories_summary.shape}")

Before summarizing: (53960, 5)
After summarizing: (7557, 5)


In this solution, we've used the color *name* inside `my_parts_categories`, but the color `id` in `hobby_train_parts`.
So we need an extra join

In [42]:
hobbytrain_and_my_parts_combined = (
    hobby_train_parts
    .merge(colors, left_on='color_id', right_on='id', suffixes=(None, '_c'))
    .merge(my_parts_categories_summary, how='left', left_on=['part_num', 'name_c'], right_on=['part_num', 'color'], suffixes=(None, '_my'))
)

In [43]:
# Fill NA values on "my" side with "0" pieces (I have zero pieces of that kind)
hobbytrain_and_my_parts_combined['quantity_my'] = hobbytrain_and_my_parts_combined['quantity_my'].fillna(0)
hobbytrain_and_my_parts_combined['quantity_excess'] = hobbytrain_and_my_parts_combined['quantity'] - hobbytrain_and_my_parts_combined['quantity_my']

df_piecesowned = hobbytrain_and_my_parts_combined.query('quantity_excess < 0')[['part_num', 'part', 'quantity_excess', 'color']]
df_piecesneeded = hobbytrain_and_my_parts_combined.query('quantity_excess > 0')[['part_num', 'part', 'quantity_excess', 'color']]
df_piecesowned['quantity_excess'] = df_piecesowned['quantity_excess'] * -1
df_piecesneeded = df_piecesneeded.rename(columns={'quantity_excess': 'quantity_missing'})
print(f"Hobby train set had {hobbytrain_and_my_parts_combined[['quantity']].sum()} parts. I am missing {df_piecesneeded['quantity_missing'].sum()} parts.")

Hobby train set had quantity    1094
dtype: int64 parts. I am missing 106.0 parts.


In [44]:
df_piecesowned.head(10)

part_num                                          part  quantity_excess  \
3        2412b  Tile Special 1 x 2 Grille with Bottom Groove            375.0   
4        2412b  Tile Special 1 x 2 Grille with Bottom Groove             27.0   
5         2420                            Plate 2 x 2 Corner            100.0   
6         2431                        Tile 1 x 4 with Groove            188.0   
7         2431                        Tile 1 x 4 with Groove            143.0   
8   2431pr0028          Tile 1 x 4 with Train Controls Print              2.0   
9         2432                Tile Special 1 x 2 with Handle              3.0   
10        2432                Tile Special 1 x 2 with Handle             44.0   
12        2436        Bracket 1 x 2 - 1 x 4 [Square Corners]             34.0   
13        2540   Plate Special 1 x 2 Side Handle [Free Ends]             57.0   

               color  
3              Black  
4                Red  
5                Red  
6              Black  
7                Red  
8   Dark Bluish Gray  
9   Dark Bluish Gray  
10             Black  
12             Black  
13             Black

In [45]:
df_piecesneeded.head(10)

part_num                                               part  \
0      2362b                                                NaN   
2       2377                             Window 1 x 2 x 2 Plane   
11      2434                                                NaN   
53      3027                                                NaN   
63     30375                                                NaN   
80      3297                                                NaN   
113     4079                                                NaN   
118    4093a  Train Base 6 x 28  with Two 1 x 2 Cutouts [3 R...   
121     4181                                                NaN   
122     4182                                                NaN   

     quantity_missing             color  
0                16.0               NaN  
2                 6.0               Red  
11                2.0               NaN  
53                2.0               NaN  
63                4.0               NaN  
80               12.0               NaN  
113               8.0               NaN  
118               1.0  Dark Bluish Gray  
121               2.0               NaN  
122               2.0               NaN

How many % of all required stones are you still missing?


In [46]:
print(f"{df_piecesneeded['quantity_missing'].sum() / hobbytrain_and_my_parts_combined['quantity'].sum() * 100:.2f} % of the pieces are missing.")

9.69 % of the pieces are missing.


What is the top 5 of the parts that you already have and the bottom 5 that you are still missing?

In [47]:
# Interpreting "top 5 parts I already have" as "top 5 parts I have the most of in excess":
df_piecesowned[['part', 'color', 'quantity_excess']].groupby(['part', 'color']).sum().sort_values('quantity_excess', ascending=False).head(5)

# Of course, you can alter this interpretation as you like to check your answers:
#
# Ignoring color
# df_piecesowned[['part', 'quantity_excess']].groupby(['part']).sum().sort_values('quantity_excess', ascending=False).head(5)


quantity_excess
part                              color                 
Brick 1 x 2                       Red             2634.0
Brick 1 x 1                       White           2538.0
                                  Red             1919.0
Plate Round 1 x 1 with Solid Stud Black           1736.0
Brick 1 x 2                       Black           1689.0

In [48]:
# Interpreting "top 5 parts I am missing" as "top 5 parts I have the most of in shortage":
df_piecesneeded[['part', 'color', 'quantity_missing']].groupby(['part', 'color']).sum().sort_values('quantity_missing', ascending=False).head(5)


quantity_missing
part                                               color                             
Glass for Window 1 x 2 x 2 Plane                   Trans-Clear                   12.0
Brick Curved 2 x 4 x 1 1/3 with Curved Top         Black                          9.0
Window 1 x 2 x 2 Plane                             Red                            6.0
Slope Curved 4 x 4 x 2/3 Triple Curved with 2 S... Dark Bluish Gray               1.0
Train Base 6 x 28  with Two 1 x 2 Cutouts [3 Ro... Dark Bluish Gray               1.0

## Bonus

You bought every set from the “Creator” theme in the year it came out. 
Make an overview per year of which parts you received.

In [49]:
all_my_parts.head(3)

id_s   name_x parent_name  set_num          name_y  year  theme_id  \
0    22  Creator     Creator  10664-1  Creative Tower  2013        22   
1    22  Creator     Creator  10664-1  Creative Tower  2013        22   
2    22  Creator     Creator  10664-1  Creative Tower  2013        22   

   num_parts  is_creator   id_i  version  inventory_id part_num  color_id  \
0       1600           1  12223        1         12223    2412b        15   
1       1600           1  12223        1         12223    2412b        71   
2       1600           1  12223        1         12223     2419         2   

   quantity is_spare  id               name     rgb is_trans  
0         4        f  15              White  FFFFFF        f  
1         4        f  71  Light Bluish Gray  A0A5A9        f  
2         2        f   2              Green  237841        f

In [50]:
# We had this one already:
# all_my_parts = (creator_sets
#     .merge(inventories, left_on='set_num', right_on='set_num', suffixes=('_s', '_i'))
#     .merge(inventory_parts, left_on='id_i', right_on='inventory_id', suffixes=(None, '_ip'))
#     # .merge(parts, left_on='part_num', right_on='part_num') # I don't think we will need parts for now
#     .merge(colors, left_on='color_id', right_on='id', suffixes=(None, '_c'))
# )
# Let's use that for this question as well:
(
    all_my_parts[['year', 'part_num', 'color_id', 'quantity']]
    [['year', 'part_num', 'quantity']]
    .groupby(['year', 'part_num'])
    .sum()
)

quantity
year part_num          
2001 14210            1
     14226c11         4
     14226c21         6
     2039             2
     2335             2
...                 ...
2021 99563           43
     99773            2
     99780           79
     99781           47
     99784           16

[10077 rows x 1 columns]

Or, if you would like it with colors:


In [51]:
(
    all_my_parts[['year', 'part_num', 'name', 'quantity']]
    .rename(columns={'name': 'color'})
    .groupby(['year', 'part_num', 'color'])
    .sum()
)

quantity
year part_num color                      
2001 14210    Green                     1
     14226c11 Black                     4
     14226c21 Black                     6
     2039     White                     2
     2335     Red                       2
...                                   ...
2021 99781    Dark Turquoise            5
              Light Bluish Gray         3
              White                     1
              Yellow                    2
     99784    Light Bluish Gray        16

[26630 rows x 1 columns]

Make a list of unique parts that only appear in one set.

In [52]:
# First try:
# In order to solve this one:
# 1. Take a list of all sets
# 2. Join with inventories
# 3. Join with inventory_parts
# 4. Join with parts
# 5. Reduce to only the columns we need (set number and part number)
# 6. Filter only on unique parts

# Steps 1-5:
sets_with_parts = (
    sets
    .merge(inventories, left_on='set_num', right_on='set_num')
    .query('version == 1')
    .merge(inventory_parts, left_on='id', right_on='inventory_id')
    .merge(colors, left_on='color_id', right_on='id', suffixes=(None, '_c'))
    .rename(columns={'name_c': 'color'})
    [['set_num', 'part_num', 'quantity', 'year', 'color']]
)
sets_with_parts.head(3)

set_num part_num  quantity  year       color
0   001-1     132a         4  1965  Light Gray
1   001-1     3020         4  1965       White
2   001-1    3062c         1  1965       White

In [53]:
# Step 6
# With the list of set / part combinations, look at the parts that are unique.
# This is a more efficient way to do it than the previous attempt.
# We will use the "sets_with_parts" table from the previous attempt.
unique_parts = sets_with_parts[['part_num', 'set_num']].groupby('part_num').count().query('set_num == 1')
unique_parts.head(3)

set_num
part_num         
003381          1
003429          1
003432          1

Which recent set(s) (2017-now) contain the most parts you still need?

In [54]:
# It helps to rephrase the question:
# When I buy a single set from 2017 or later, what is the single set that leaves me with the least missing parts after buying the set?
print(f"Before I buy any new set, the number of pieces I'm missing is: {df_piecesneeded['quantity_missing'].sum()}")


Before I buy any new set, the number of pieces I'm missing is: 106.0


## Wrong solution 1: counting how much pieces you need

Check if you can see what's wrong here:

In [55]:

# 1. Filter only 2017+
sets_from_2017 = sets_with_parts.query('year > 2017')
# 2. Join with parts I need
df_piecesneeded_afterbuying = df_piecesneeded.merge(sets_from_2017, how='left', left_on=['part_num', 'color'], right_on=['part_num', 'color'], suffixes=(None, '_s'))
# 3. Calculate how many pieces are missing afterwards
df_piecesneeded_afterbuying['quantity_missing_afterbuying'] = df_piecesneeded_afterbuying['quantity_missing'] - df_piecesneeded_afterbuying['quantity']
df_piecesneeded_afterbuying['quantity_missing_afterbuying'] = df_piecesneeded_afterbuying['quantity_missing_afterbuying'].clip(lower=0)

In [56]:
# 4. Group by set_num, sum the missing pieces
df_pnab_grouped = df_piecesneeded_afterbuying[['set_num', 'quantity_missing_afterbuying']].groupby('set_num').sum().reset_index()

In [57]:
# 5. Order by missing pieces
df_pnab_grouped[['set_num', 'quantity_missing_afterbuying']].groupby('set_num').sum().sort_values('quantity_missing_afterbuying', ascending=True).head(5)

quantity_missing_afterbuying
set_num                              
75258-1                           0.0
21323-1                           1.0
10274-1                           4.0
43115-1                           5.0
31085-1                           7.0

Now have a look at the LEGO website: set #75258 turns out on top, but is not very likely to contain all missing pieces I need for a red brick LEGO train...
[75258](https://www.lego.com/en-us/product/anakin-s-podracer-20th-anniversary-edition-75258).

So what went wrong?

* Only the missing parts that are available in the set we're buying show up. 
* So we don't have the number of *really* missing tiles:
* For example, the top 5 missing pieces are:

In [58]:
df_piecesneeded.sort_values('quantity_missing', ascending=False).head(5)

part_num                                        part  quantity_missing  \
0      2362b                                         NaN              16.0   
138     4862            Glass for Window 1 x 2 x 2 Plane              12.0   
80      3297                                         NaN              12.0   
151     6081  Brick Curved 2 x 4 x 1 1/3 with Curved Top               9.0   
113     4079                                         NaN               8.0   

           color  
0            NaN  
138  Trans-Clear  
80           NaN  
151        Black  
113          NaN

But in the `75258-1` set only a SINGLE tile can be found that we're missing:

In [59]:
df_piecesneeded_afterbuying.query('set_num=="75258-1"')

part_num                                               part  \
15    45677  Slope Curved 4 x 4 x 2/3 Triple Curved with 2 ...   

    quantity_missing             color  set_num  quantity    year  \
15               1.0  Dark Bluish Gray  75258-1       1.0  2019.0   

    quantity_missing_afterbuying  
15                           0.0

Altering your merge (join) type won't help here: 

* it only assures that all missing tiles are visible in the dataset
* it doesn't assure that all missing tiles are visible with each LEGO set
  * Part 2362b for example is in no LEGO set after 2017:
  * So although we're missing 16 of these pieces, as long as we simply group on LEGO set this one simply won't show up.

In [60]:
df_piecesneeded_afterbuying.query('part_num=="2362b"')

part_num part  quantity_missing color set_num  quantity  year  \
0    2362b  NaN              16.0   NaN     NaN       NaN   NaN   

   quantity_missing_afterbuying  
0                           NaN

To visualize this, let's assume we are missing only three pieces:

| Missing piece | # missing |
|---------------|-----------|
| Red brick     | 1         |
| Black brick   | 25        |
| Yellow brick  | 13        |

We also have a list of two sets containing pieces:

| Set # | Piece       | # of pieces in set |
|-------|-------------|--------------------|
| Set 1 | Red brick   | 25                 |
| Set 2 | Red brick   | 3                  |
| Set 2 | Black brick | 3                  |

Now when we do a **left** join between the tables on the piece, the following table is generated:

| Set # | Piece        | # missing | # of pieces in set |
|-------|--------------|-----------|--------------------|
| Set 1 | Red brick    | 1         | 25                 |
| Set 2 | Red brick    | 1         | 3                  |
| Set 2 | Black brick  | 25        | 3                  |
|       | Yellow Brick | 13        |                    |

There are two problems here:

1. As you can see, the **yellow** brick from this example is not linked to any set
2. Although we could possibly add the "non-linked" bricks to all sets, this doesn't solve everything: The black brick still ends up in "set 2", but not in "set 1"


## Solution

In order to make this work, we need to rethink the order in which we're doing our merges/joins:

1. First, we need to have a list of missing parts (which is `df_piecesneeded` already)
   | Missing piece | # missing |
   |---------------|-----------|
   | Red brick     | 1         |
   | Black brick   | 25        |
   | Yellow brick  | 13        |
2. Then we need to do a CROSS JOIN with all sets (only sets, no parts yet!) from 2017
   * This gives a cartesian product
   * We then end up with every piece we need, listed within every set after 2017
   | Set # | Piece        | # missing |
   |-------|--------------|-----------|
   | Set 1 | Red brick    | 1         |
   | Set 1 | Black brick  | 25        |
   | Set 1 | Yellow brick | 13        |
   | Set 2 | Red brick    | 1         |
   | Set 2 | Black brick  | 25        |
   | Set 2 | Yellow brick | 13        |

3. Only **after** this cross join we will do a left join towards the inventory and parts that are actually in the set
   | Set # | Piece        | # missing | # of pieces in set |
   |-------|--------------|-----------|--------------------|
   | Set 1 | Red brick    | 1         | 25                 |
   | Set 1 | Black brick  | 25        |                    |
   | Set 1 | Yellow Brick | 13        |                    |
   | Set 2 | Red brick    | 1         | 3                  |
   | Set 2 | Black brick  | 25        | 3                  |
   | Set 2 | Yellow Brick | 13        |                    |

... and now we can calculate by what degree the new sets satisfy our need for parts.

In [70]:
# 1. We have already: df_piecesneeded
# 2. Cross join the missing parts with all sets from 2017
# Note that the difference with an earlier attempt is that we DON'T yet join the parts
sets_from_2017 = sets.query('year > 2017')
df_piecesneeded_sets2017 = df_piecesneeded.merge(sets_from_2017, how='cross', suffixes=(None, '_s'))
# 3. Now join with inventories and parts
df_piecesneeded_inventories_2017 = df_piecesneeded_sets2017.merge(inventories, how='left', left_on='set_num', right_on='set_num').query('version == 1')
df_piecesneeded_parts_2017 = df_piecesneeded_inventories_2017.merge(inventory_parts, how='left', left_on=['id', 'part_num'], right_on=['inventory_id', 'part_num'])
# 4. For sets where the missing parts are not in the set, we will have NaN values. We will fill these with 0.
df_piecesneeded_parts_2017['quantity'] = df_piecesneeded_parts_2017['quantity'].fillna(0)
# 5. Calculate the missing parts after buying the set
df_piecesneeded_parts_2017['quantity_missing_afterbuying'] = df_piecesneeded_parts_2017['quantity_missing'] - df_piecesneeded_parts_2017['quantity']
df_piecesneeded_parts_2017['quantity_missing_afterbuying'] = df_piecesneeded_parts_2017['quantity_missing_afterbuying'].clip(lower=0)
# 6. Group by set_num, sum the missing parts
df_pnab_grouped_2017 = df_piecesneeded_parts_2017[['set_num', 'quantity_missing_afterbuying']].groupby('set_num').sum().reset_index()
# 7. Order by missing pieces
df_pnab_grouped_2017.sort_values('quantity_missing_afterbuying', ascending=True).head(5)

set_num  quantity_missing_afterbuying
305   21323-1                          89.0
1691  60266-1                          89.0
2600  80101-1                          94.0
25    10283-1                          96.0
1617  60188-1                          97.0

Note that low numbers of 'quantity missing' sets are likely to have an error: it is not very plausible to have a range of 106 different pieces to be found in a single set. So the quantity missing after buying a new set is likely to be still high.

Is this set still available? Answer this last question by scraping the LEGO website (tip: BeautifulSoup)

In [72]:
import requests
from bs4 import BeautifulSoup

# Define the URL of the LEGO set page
url = "https://www.lego.com/en-us/product/21323"

# Send a GET request to the URL
response = requests.get(url)

# Create a BeautifulSoup object from the response content
soup = BeautifulSoup(response.content, "html.parser")

In [73]:
# Check if the set exists by looking for the product title element
product_availability = soup.find("p", attrs={"data-test": "product-overview-availability"})
product_availability.text == 'Available now'

True

Do you know any alternative ways to answer the main question? If so, please explain how to do this.